Autotest api

agaltsoff opened the pull request:
**Реализация** #753 

**Сделанные доработки**

Реализован минимальный набор методов API автоматического тестирования и утверждений, позволяющий оценить синтаксис законченного теста.

*Утверждения в УтвержденияBDD*

ПодключениеКлиентаТестирования
ОтключениеКлиентаТестирования
ОткрытиеФормыСписка
ПоискВСписке
УстановкаОтбораВСписке
ОткрытиеВСписке
КоличествоСтрокВСписке
Проведение

**Описание**

Методы для работы с механизмом автоматизированого тестирования содержатся в ТестКлиенты.
УтвержденияBDD использует эти методы в утверждениях автоматизированого тестирования.

ТестКлиенты сохраняет значение окна и элемента, которые были получены при выполнении каждого метода. Последующие действия выполняются относительно этих значений. Эти значения можно получить с помощью соответствующих методов интерфейса ТестКлиенты.

Открытие окон в ТестКлиенты это процедуры, поиск значений - булевые функции

ТестКлиенты выполняет действия без проверок, все проверки выполняются в УтвержденияBDD.

**Пример теста**

```
    Ожидаем.ПодключениеКлиентаТестирования("Администратор", "");

    Ожидаем
        .ОткрытиеФормыСписка("Документ.Документ1", "Документ1")
        .ПоискВСписке(Новый Структура("Номер,Дата", "000000002", "26.11.2017 23:13:37"))
        .ОткрытиеВСписке("Документ1 000000002 от 26.11.2017 23:13:37")
        .Проведение()
    ;

    Ожидаем
        .ОткрытиеФормыСписка("РегистрСведений.РСПодчиненныйРегистратору", "РСПодчиненный регистратору")
        .УстановкаОтбораВСписке(Новый Структура("Регистратор", "Документ1 000000002 от 26.11.2017 23:13:37"))
        .КоличествоСтрокВСписке("Движения документа Документ1 000000002 от 26.11.2017 23:13:37").Равно(1);
    ;

    Ожидаем.ОтключениеКлиентаТестирования();
```

agaltsoff committed:
Шаболн плагина УтвержденияBDDForms и тестов к нему

agaltsoff committed:
УтвержденияBDDForms ПодключениеКлиентаТестирования ОтключениеКлиентаТестирования плюс тесты на это

agaltsoff committed:
Перенёс всё из формы УтвержденияBDDForms в УтвержденияBDD, удалил УтвержденияBDDForms

agaltsoff committed:
УтвержденияBDD работа с клиентом тестирования через плагин ТестКлиенты
В тесте только подключается и отключается клиент, тест проходит но приложение клиента не закрывается, потому что отключение вызывается раньше чем откроется окно клиента.

agaltsoff committed:
Открытие формы списка. Функционал действует но форма не открывается))

agaltsoff committed:
Исправил запуск и завершение в ТестКлиенты, тест открытия формы в Тесты_УтвержденияBDDForms. Проходят тесты подключение, открытие списка, отключение.

agaltsoff committed:
Поиск в списке и открытие в ТестКлиенты.

agaltsoff committed:
Выделил области с помощью #Область #КонецОбласти

agaltsoff committed:
Подключение, Открытие формы списка, Поиск в списке, Открытие в списке текущей строки, Отключение.

Методы в ТестКлиенты и УтвержденияBDD плюс тесты в Тест_УтвержденияBDDForms. Более-менее устоялась идеология.

agaltsoff committed:
Убрал сообщения перед запуском и после запуска теста, добавил тест вызова уверждений через точку.

agaltsoff committed:
Утверждения ОткрытиеВСписке, УстановкаОтбораВСписке, КоличествоСтрокВСписке, Проведение

C их помощью реализован тест проведения документа и проверки количества его движений по регистру в Тест_УтвержденияBDDForms.
Тест выполняется на данных \Tests\1Cv8_xUnitFor1CForms.dt.
Конфигурация там из тестовой базы, в один документ добавлена обработка проведения.

agaltsoff committed:
Тесты_УтвержденияBDDForms выполняются на пустой тесовой конфигурации без изменений. Тестовые данные создаются и удаляются в тестах.

Добавил cf тесовой конфигурации из ветки master в \Tests.
В модуль формы Тесты_УтвержденияBDDForms добавил процедуры для работы с таблицей параметров на форме. На сервере записываем, на клиенте читаем.
Хотел  использовать для передачи параметров между тестами но потом сделал передачу через контекст ядра.

agaltsoff committed:
Версия 4.2.0.3

agaltsoff committed:
Отбор в списке по регистратору. Отладка.

artbear commented:
Хорошее дополнение, 

но мне не нравится, что смешаны 2-е функциональности - Утверждения и работа с тест-клиентом

Предлагаю разделить эту функциональность.
Например, 
+ сделать отдельный плагин `УправлениеТестКлиентом`, в котором и реализовать нужное API
  + если нравится другое название, предлагайте
+ пример кода
```
ТестКлиент.ПодключениеКлиентаТестирования("Администратор", "");

    ТестКлиент
        .ОткрытиеФормыСписка("Документ.Документ1", "Документ1")
        .ПоискВСписке(Новый Структура("Номер,Дата", "000000002", "26.11.2017 23:13:37"))
        .ОткрытиеВСписке("Документ1 000000002 от 26.11.2017 23:13:37")
        .Проведение()
    ;

    ТестКлиент
        .ОткрытиеФормыСписка("РегистрСведений.РСПодчиненныйРегистратору", "РСПодчиненный регистратору")
        .УстановкаОтбораВСписке(Новый Структура("Регистратор", "Документ1 000000002 от 26.11.2017 23:13:37"))
        .КоличествоСтрокВСписке("Движения документа Документ1 000000002 от 26.11.2017 23:13:37").Равно(1);
    ;

    Ожидаем.ОтключениеКлиентаТестирования();
```

или предложите другие варианты

@agaltsoff @VladFrost @nixel2007 @EvilBeaver что скажете?

VladFrost commented on src/Plugins/ТестКлиенты/Form/Форма/Форма.bsl:
```diff
@@ -144,12 +162,227 @@
+	
+	УбедитьсяЧтоЭтоСпискок();
+	
+	ТестируемыйЭлемент.ВыделитьВсеСтроки();
```
Здесь опасность: если в списке больше 1000 строк, платформа выведет предупреждение и выполнение будет приостановлено, пока пользователь вручную не закроет предупреждение.

agaltsoff replied:
Вот да, надо придумать как количество строк в таблице получать.
Нормального метода не нашёл.

agaltsoff replied:
Тут похоже концептуальная проблема невозможности определения количества строк в динамическом списке разумными средствами.
Сделал проверку на пустой список потому что по сути нужна только она.

kuntashov replied:
Все же странно в сценарных тестах увидеть такой объем данных в списках.
Про данную особенность помнить надо, но решать ее сейчас особого смысла нет.

VladFrost replied:
@kuntashov как вариант, если тесты прогоняются на типовой демобазе: там вполне может быть несколько тысяч строк в списках.

VladFrost commented on src/Plugins/ТестКлиенты/Form/Форма/Форма.bsl:
```diff
@@ -290,4 +520,132 @@
-// } Helpers
\ No newline at end of file
+&НаКлиенте
+Функция ЗабытьЗапущенныйКлиент(ИмяПользователя, Порт)
```
Про эту функцию я не понял, как она должна использоваться.

agaltsoff replied:
Удалять клиента из списка после отключения.

VladFrost replied:
Это да. Но она не экспортная, и вызовов её в этом модуле нет.

agaltsoff replied:
Хотел использовать, но как-то обошлось без неё потом. Удалю.

agaltsoff replied:
Удалил.

VladFrost commented on src/Plugins/УтвержденияBDD/Form/Форма/Форма.bsl:
```diff
@@ -205,4 +222,253 @@
+#Область ИнтерфейсАвтоматическогоТестирования
+
+&НаКлиенте
+Функция ПодключениеКлиентаТестирования(ИмяПользователя = "", Пароль = "", Порт = 1539) Экспорт
```
Почему номер порта стал 1539? По-умолчанию же 1538.

agaltsoff replied:
1538 менеджер тестирования использует по умолчанию.

VladFrost replied:
И почему здесь нужен 1539?

agaltsoff replied:
У меня не получалось найти клиент запущенный на одном порту с менеджером. Возможно причина была в другом. Сейчас проверил, так работает.
Вернул 1538.

agaltsoff replied:
Вот такие ошибки теперь периодически появляются:

`
{Обработка.ТестКлиенты.Форма.Форма.Форма(106)}: Не смогли установить соединение с тестовым приложением для пользователя Администратор!
{Обработка.ТестКлиенты.Форма.Форма.Форма(97)}: Ошибка при вызове метода контекста (УстановитьСоединение): Ошибка подключения к клиенту тестирования. Менеджер не может управлять сам собой.: Ошибка подключения к клиенту тестирования. Менеджер не может управлять сам собой.
 `

VladFrost replied:
А какая версия платформы используется? Такое подозрение, будто клиент тестирования запускается одновременно с ключами /TESTMANAGER и /TESTCLIENT

agaltsoff replied:
Платформа 8.3.10.2580
Такое возможно и было, у меня параметрах базы в Стартере был ключ /TESTCLIENT с которым был запущен Конфигуратор и он видимо с ним же и запустил отладку и плюс менеджер тестирования из параметров запуска Предприятия в самом Конфигураторе . Сейчас я убрал оттуда этот ключ, поставлю везде порт 1538, понаблюдаю.
Но как-то это всё равно старнно иметь два процесса на одном порту.

VladFrost replied:
Дык порт слушает только один процесс - тот, который запущен с ключом `/TESTCLIENT`.
Процесс, запущенный c ключом `/TESTMANAGER`, порт не слушает, а только подключается к тому, который запущен с ключом `/TESTCLIENT`.

VladFrost replied:
Вообще, для отладки, с какими ключами был процесс запущен, я использую Process Explorer:
![2017-11-30 17_30_18](https://user-images.githubusercontent.com/11960251/33435816-3b8e2d8c-d5f4-11e7-98c5-25abd82e2019.png)

agaltsoff replied:
Понятно. Думал менеджер тоже слушает этот порт, а подключается к тому который укажешь в Новый ТестируемоеПриложение. 
Тогда оставлю как было.

agaltsoff replied:
Там ещё веселее всё. Конфигуратор при запуске Предприятия берёт параметры из строки Стартера даже те, которые были туда добавлены **после** запуска Конфигуратора.
Т.е. я запустил конфигуратор, потом добавил /TESTCLIENT чтобы запустить эту базу вручную в режиме тестирования и после этого уже запустил Предприятие как менеджер тестирования из Конфигуратора.

VladFrost replied:
Для запуска из конфигуратора лучше здесь рулить:
![2017-12-01 09_12_25](https://user-images.githubusercontent.com/11960251/33470163-e392972e-d677-11e7-856a-9ae8d582134d.png)

agaltsoff replied:
Да я так и делаю. Просто после того как конфигуратор запустил понадобилось эту же базу запустить как тест клиент. В стартере поставил параметр в строке запуска, запустил, потом через какое-то время запустил предприятие из конфигуратора. И он взял этот параметр из стартера, хотя на момент запуска конфигуратора флага ещё не было. 
И вообще, почему конфигуратор берёт параметры предприятия из списка баз из строки с которой он был запущен? Странно...
Сейчас просто сделал отдельную строку для запуска базы в режиме клиента тестирования чтобы не путаться.

VladFrost replied:
А зачем его отдельно запускать? Плагин же сам его запускает: https://github.com/xDrivenDevelopment/xUnitFor1C/blob/develop/src/Plugins/%D0%A2%D0%B5%D1%81%D1%82%D0%9A%D0%BB%D0%B8%D0%B5%D0%BD%D1%82%D1%8B/Form/%D0%A4%D0%BE%D1%80%D0%BC%D0%B0/%D0%A4%D0%BE%D1%80%D0%BC%D0%B0.bsl#L77

agaltsoff replied:
Что-то нужно было проверить руками, разовый случай.

VladFrost commented on src/Plugins/ТестКлиенты/Form/Форма/Форма.bsl:
```diff
@@ -290,4 +520,132 @@
+
+
+#КонецОбласти 
+
```
Пустые строки

agaltsoff replied:
Чтобы последнюю строку с текстом можно было поместить в середину монитора при редактировании.

VladFrost commented on src/Plugins/ТестКлиенты/Form/Форма/Форма.bsl:
```diff
@@ -144,12 +162,227 @@
+		
+		Если ИнформацияОбОшибке <> Неопределено Тогда
+			ОписаниеОшибки= 
+			    НСтр("ru='Описание=';en='Description='") + ИнформацияОбОшибке.Описание + "'" + Символы.ПС +
```
Здесь хочется видеть СтрШаблон()

agaltsoff replied:
Это из синтаксис-помощника)) 
Исправлю.

VladFrost commented on src/Plugins/ТестКлиенты/Form/Форма/Форма.bsl:
```diff
@@ -290,4 +520,132 @@
+КонецПроцедуры
+
+&НаКлиенте
+Процедура УбедитьсяЧтоЭтоСпискок()
```
Опечатка в имени процедуры

agaltsoff replied:
Исправлю))

agaltsoff replied:
Исправлю.

VladFrost commented on src/Plugins/УтвержденияBDD/Form/Форма/Форма.bsl:
```diff
@@ -205,4 +222,253 @@
+КонецФункции
+
+&НаКлиенте
+Функция ПоискВСписке(СтруктураПоиска) Экспорт
```
Чем "ПоискВСписке" отличается от "НайтиВСписке"?

agaltsoff replied:
УтвержденияBDD.ПоискВСписке вызывает ТестКлиенты.НайтиВСписке, анализирует результат и при возникновении ошибки вызывает ВызватьОшибкуПроверки.
ТестКлиенты.НайтиВСписке непосредственно выполняет поиск в списке через механизм автотестирования.
Т.е. ТестКлиенты это low level, УтвержденияBDD это high level.
Названия методов в утверждениях выбраны так чтобы они сочетались с Ожидаем.

VladFrost commented on src/Plugins/УтвержденияBDD/Form/Форма/Форма.bsl:
```diff
@@ -205,4 +222,253 @@
+СтатусыРезультатаТестирования = Новый ФиксированнаяСтруктура(СтатусыРезультатаТестирования);
+
+#КонецОбласти 
+
```
Пустые строки

agaltsoff replied:
Чтобы последнюю строку с текстом можно было поместить в середину монитора при редактировании.

agaltsoff commented:
@artbear

Внутри так всё и происходит. УтвержденияBDD при инициализации загружает ТестКлиенты и вся работа с механизмом автоматического тестирования идёт через него.

Например Ожидаем.ОткрытиеВСписке вызывает ТестКлиенты.ОткрытьВСписке, анализирует результат и в случае ошибки вызывает стандартное исключение которое потом красиво выводится.

Можно в модуле теста сделать глобальную переменную и при инициализации теста записать труда ссылку на ТестКлиенты загруженный в УтвержденияBDD. Тогда можно будет вызывать методы, больше относящиеся к механизму автотестирования чем к функционалу самого теста напрямую. Только нужно определится какие.

Например то же ОткрытиеВСписке вроде бы относится к автотестированию, но если мы проверяем настройки прав доступа, это уже и сам тест.

artbear commented:
Внутри может быть все что угодно.
Мы же говорим об API, которое в дальнейшем будут юзать клиенты продукта.
Я **хочу развязать API Утверждений и API ТестКлиента** и не смешивать.

Пусть методы API ТестКлиента внутри себя вызывают необходимые утверждения и выбрасывают красивые ошибки.

Общая схема работы - проще и удобнее через ТестКлиент, а не через Ожидаем.
Пример кода вызова я приводил.

artbear commented:
Методы API должны помочь клиенту продукта. 
Например, в коде метода может быть инкапсулировано общее для разных тестов поведение.
Мы об этом в гиттере вроде бы уже говорили.

artbear commented:
Я не хочу постоянно дорабатывать Утверждения.
Их цель - простые утверждения.

Более сложные утверждения д.б. реализованы в других плагинах.
Например, так уже сделано в ЗапросыБДД, Сериализатор и других.
ТестКлиент должен быть сделан по этой же схеме :)

Тем более, что самое сложное (методы) Вы уже реализовали.
Теперь их нужно просто перенести в новый плагин.

agaltsoff commented:
Понятно. Так даже проще будет. Из Утверждений работу с автотестированием убираем в отдельный плагин для Утверждений автотестирования?

artbear commented:
Только предлагаю новый плагин все-таки назвать "ТестКлиент", а не "Утверждения автотестирования" :)

agaltsoff commented:
Так есть же уже один ТестКлиент**ы**. Путаницы не будет?

agaltsoff commented:
Ещё есть УтвержденияПроверкаТаблиц. Видимо утверждения для специальных случаев. На мой взгляд по аналогии будет УтвержденияBDDАвтотест. BDD потому что *ОткрытиеВСписке* сочетается с *Ожидаем*. Без BDD тогда *Открыть...*

Однако как скажете. Вы же здесь [BDFL](https://en.m.wikipedia.org/wiki/Benevolent_Dictator_For_Life) ))

artbear commented:
Все-таки мне больше нравится 
+ `ТестКлиент` - для работы с одним тест-клиентом
+ `ТестКлиенты`- управление тест-клиентами
  + похоже, пока не выпустили релиз, лучше назвать `МенеджерТестКлиентов`

artbear commented:
Честно, УтвержденияBDDАвтотест - жесть. Длинно, странно, невнятно

Потом все время по коду удивляться над 
`УтвержденияBDDАвтотест.ОткрытиеФормыСписка(....)` :(

agaltsoff commented:
Не ну в коде писать такое конечно не стоит. УтвержденияBDD трансформируются в Ожидаем, здесь было бы ОжидаемАвтотест.

agaltsoff commented:
В результате
- ТестКлиенты переименовать в МенеджерТестКлиентов
- API автотестирования перенести из УтвержденияBDD в новый плагин ТестКлиент?

artbear commented:
Да, я это и предлагаю.

Единственное - @VladFrost что скажешь по `ТестКлиенты переименовать в МенеджерТестКлиентов` ?

agaltsoff committed:
???????????.???????????????????????????????????? ???? ????. ????????.

agaltsoff committed:
Рефакторинг УстановитьОтборВСпискеПоРегистратору с функциями Найти...

VladFrost commented:
`МенеджерТестКлиентов` подходит

agaltsoff committed:
ТестКлиенты отбор в списке по регистратору

agaltsoff committed:
ТестКлиенты.КОписаниюСтроки добавил тип Строка и Неопределено

agaltsoff committed:
Тест для установки отбора в списке, не проходит.

agaltsoff committed:
ТестКлиенты.УстановитьОтборВСпискеПоРегистратору плюс тест. Работает.

ПерейтиКСтроке на пустом списке вызывает ошибку. Добавил везде перед вызовом этой функции проверку количества строк в списке.

agaltsoff committed:
ЭтоПустойСписок без использования выделения всех строк.

agaltsoff committed:
УстановитьОтборВСпискеПоРегистратору добавил проверки на открытие окон, когда они не проходят все предыдущие окна закрываются

agaltsoff committed:
УстановитьОтборВСписке с помощью расширенного поиска с учётом возможного выбора типа данных

artbear commented:
>В результате
- ТестКлиенты переименовать в МенеджерТестКлиентов
- API автотестирования перенести из УтвержденияBDD в новый плагин ТестКлиент

@agaltsoff Ждем доработки

agaltsoff committed:
Отладка через БСП и ИР. Работает.

agaltsoff committed:
Отладка через БСП и ИР работает в рабочей базе.

Обновил ИР до 4.37

agaltsoff committed:
minor ТестКлиенты скопировал сюда из сетевого каталога. Сообщение об ошибке теста с новой строки.

agaltsoff committed:
Плагины и тесты из дополнительных отчётов и обработки БСП загружаются. Проверено. Отладка работает через раз.

agaltsoff committed:
В xddTestRunner перенёс загрузку плагинов из дополнительных обработок БСП в инициализацию.

agaltsoff committed:
Сделал вызовы методов БСП через Выполнить/Вычислить для запуска без ошибок при отсутствии БСП.

agaltsoff committed:
Добавил модуль ТестКлиент, работу с клиентом тестирования перенёс туда, ТестКлиенты переименовал в МенеджерТестКлиентов

agaltsoff committed:
Все процедуры перенёс. Тест начал выполняться.

В xddTestRunner инициализацию плагинов сделал после загрузки всех плагинов чтобы плагину в инициализации был доступен любой другой плагин.

agaltsoff committed:
Перенёс всё в конфигурацию.

agaltsoff committed:
Тесты ТестКлиент проходят.

agaltsoff committed:
Убрал неиспользуемое из Тест_ТестКлиент

agaltsoff committed:
Выгрузил тестовую конфигурацию с новыми плагинами

agaltsoff commented:
> В результате
>- ТестКлиенты переименовать в МенеджерТестКлиентов
>- API автотестирования перенести из УтвержденияBDD в новый плагин ТестКлиент
>
>@agaltsoff Ждем доработки

@artbear Сделал.

artbear commented:
@agaltsoff Очень много изменений/коммитов.
Я несколько запутался :( - есть изменения в конфигурации, есть в файлах и т.п.

Можешь достаточно коротко описать финальное состояние ?

agaltsoff committed:
В xddTestRunner вернул загрузку извнешних обработок БСП в ПодключитьПлагиныНаСервере

Поисправлял мелкие ошибки менеджере тест клиентов. Ошибки ещё остались.
В ТестКлиент заменил старое название плагина ТестКлиенты на МенеджерТестКлиентов, как название файла плагина.

Отладка по HTTP внешних обработок БСП заработала после перерегистрации библиотек ИР.
Наимнование обработки в справочнике ДОЛЖНО СОВПАДАТЬ С ИМЕНЕМ ФАЙЛА обработки.

agaltsoff committed:
Исправил ошибки в менеджере тест клиентов при установке отбора в списке через расширенный поиск.

В поле даты при выборе открывалось окно календаря вместо окна выбора типа данных и текущее окно становилось неопределено.
В поле строки окно выбора типа данных не открывалось, но проверка выдавала что оно открыто.
Раньше вроде такого не было. Смена платформы повлияла?

Отладка вызова методов управляемых форм без их открытия через внешние обработки БСП по HTTP пока работает.

agaltsoff committed:
Исправил мелкие ошибки. Переименовал области по стандартам 1С.

agaltsoff committed:
Исправил мелкие ошибки. Тест в рабочей базе теоретически проходит.

Не работает установка отбора перед поиском в списке чтобы поиск выполнялся быстрее.
Не работает вызов метода Утверждений BDD в тест клиенте потому что они неопределены хотя инициализация объекта плагина выполняется.

Без этих двух ошибок тесты проходят.

agaltsoff committed:
Исправил установку отбора по параметрам поиска перед поиском в списке.

При активизации Что искать для простого типа данных окно списка становится активным окном приложения и оно закрыввалось вместо окна списка выбора, которое открывается для ссылочного типа.

agaltsoff committed:
Исправил вызов утверждений BDD в тест клиенте. Тест в рабочей базе проходит.

Утверждения BDD вызывались у объекта обработки ТестКлиент, а он не был инициализирован.
Пробовал инициализировать объекты форм плагинов при получении форм при тестировании на клиенте, но не до конца разобрался с типами.
Сделал использование утверждений в тест клиенте BDD через форму плагина УтвержденияBDD как в-общем то и надо было сделать изначально.

Вопрос, нужно ли инициализировать объекты форм плагинов при использовании плагинов в виде форм на клиенте и как это сделать.

agaltsoff commented:
@artbear Да там вроде кроме

>ТестКлиенты переименовать в МенеджерТестКлиентов
API автотестирования перенести из УтвержденияBDD в новый плагин ТестКлиент

должны быть только изменения для тестирования через внешние обработки БСП.

Я тогда посмотрю по [diffview](https://github.com/xDrivenDevelopment/xUnitFor1C/compare/develop...agaltsoff:autotest_api) какие ещё bsl файлы менялись кроме _ТестКлиент_ _МенеджерТестКлиентов_ _УтвержденияBDD_ из тех что остались и сделаю описание изменений.

artbear commented:
@agaltsoff Ага, жду описания

agaltsoff commented:
В итоге сделал branch с того места где мы договорились

>- ТестКлиенты переименовать в МенеджерТестКлиентов
>- API автотестирования перенести из УтвержденияBDD в новый плагин ТестКлиент

и скопировал туда конечный результат всех последующих изменений. Вроде получилось то что надо. Видны только результаты изменений без промежуточных шагов. Назвал [autotest](https://github.com/agaltsoff/xUnitFor1C/tree/autotest). Из него сделал [pull request](https://github.com/xDrivenDevelopment/xUnitFor1C/pull/776), описание там, а этот закрою.

Пробовал конечно join commit и merge, но не взлетело. Видимо не до конца разобрался с этим ещё.
